### PR TITLE
Introduce exit! command

### PR DESCRIPTION
**Pull request by ignacio-chiazzo**
# Problem
Currently, `exit!` does not save the history of the console' input like the following example

```console
irb(main):001:0> a = 1
=> 1
irb(main):002:0> b = 2
=> 2
irb(main):003:0> exit
```

Open another session and type up-arrow to check the above history
`irb(main):001:0> b = 2 # (history is saved!)`

If we run the same for `exit!`, the history isn't saved.
Closes [#612](https://github.com/ruby/irb/issues/612)

# Solution

I extended the command `exit!` and used a new class `ExitForcedAction`. Note that we can't use `Exit!` as the class name since `!` isn't allowed.

This will call the `irb_exit!` method, which will throw the same exception as exit (`throw :IRB_EXIT`) and the main method `run` catches it and calls `save_history``.

# Demo

In the following example, I test that the history is saved and `Kernel.exit!` exits the console even when the session is nested.

https://github.com/ruby/irb/assets/11672878/f6993f51-e729-4f4c-9ce4-b40a9137c925

### CODE REVIEW

**Comment by st0012** [comment]
Calling `exit!` now behaves differently then `exit`:

**exit**

```
$ be ruby test.rb

From: test.rb @ line 1 :

 => 1: binding.irb
    2: binding.irb

irb(main):001> exit

From: test.rb @ line 2 :

    1: binding.irb
 => 2: binding.irb

irb(main):001> exit
```

**exit!**

```
$ be ruby test.rb

From: test.rb @ line 1 :

 => 1: binding.irb
    2: binding.irb

irb(main):001> exit! # the process exists right away
```

I think the solution should also maintain this behaviour while saving the history.

File: lib/irb.rb [hunk 1]
```diff
@@ -889,6 +889,10 @@ def IRB.irb_exit(*)
     throw :IRB_EXIT
   end
 
+  def IRB.irb_exit!(*)
+    throw :IRB_EXIT!
```
Comment (st0012) [inline lib/irb.rb:893]: We can pass return value when throwing. So if we do `throw :IRB_EXIT, true` here, I think `Irb#run` can be simplified to:

```rb
      # ....
      force_quit = false

      begin
        force_quit = catch(:IRB_EXIT) do
          eval_input
        end
      ensure
        trap("SIGINT", prev_trap)
        conf[:AT_EXIT].each{|hook| hook.call}
        context.io.save_history if save_history
        Kernel.exit!(0) if force_quit
      end
```

File: lib/irb/cmd/exit_forced_action.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  module ExtendCommand
+    class ExitForcedAction < Nop
+      category "IRB"
+      description "Exit the current irb session with exit!."
+
+      def execute(*)
+        IRB.irb_exit!
+      rescue UncaughtThrowError
+        Kernel.exit
```
Comment (st0012) [inline lib/irb/cmd/exit_forced_action.rb:16]: Maybe this should be `Kernel.exit!(0)` instead?

File: lib/irb.rb [hunk 3]
```diff
@@ -979,13 +984,21 @@ def run(conf = IRB.conf)
       end
 
       begin
-        catch(:IRB_EXIT) do
+        @forced_exit = false
```
Comment (st0012) [inline lib/irb.rb:987]: Why does it have to be an ivar instead of a local?

File: lib/irb.rb [hunk 4]
```diff
@@ -968,7 +972,8 @@ def run(conf = IRB.conf)
       conf[:IRB_RC].call(context) if conf[:IRB_RC]
       conf[:MAIN_CONTEXT] = context
 
-      save_history = !in_nested_session && conf[:SAVE_HISTORY] && context.io.support_history_saving?
+      supports_history_saving = conf[:SAVE_HISTORY] && context.io.support_history_saving?
+      save_history = !in_nested_session && supports_history_saving
```
Comment (st0012) [inline lib/irb.rb:976]: Why do we need to separate this?

File: lib/irb/cmd/exit_forced_action.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+require_relative "nop"
+
+module IRB
+  # :stopdoc:
+
+  module ExtendCommand
+    class ExitForcedAction < Nop
+      category "IRB"
+      description "Exit the current irb session with exit!."
```
Comment (st0012) [inline lib/irb/cmd/exit_forced_action.rb:11]: I think we can make the effect of this command more clear:

```suggestion
      description "Exit the current process."
```

**Review by st0012 — COMMENTED** [review]
I think we're pretty close, just want to clarify 2 changes.

File: lib/irb.rb (hunk 4 above)
Comment (ignacio-chiazzo) [reply lib/irb.rb:976]: `save_history` is false when `in_nested_session` is true. When calling `exit!`, we still want to save the history even if the session is nested.

```console
binding.irb
binding.irb
exit! -> should save history
```

```console
binding.irb
binding.irb
exit -> should not save history
a = 3
b = 44
exit -> should save history
```

File: lib/irb.rb (hunk 3 above)
Comment (ignacio-chiazzo) [reply lib/irb.rb]: not need to. I just updated this part.

File: lib/irb.rb (hunk 4 above)
Comment (st0012) [reply lib/irb.rb:976]: I see, that's a good catch 👍 
Do you mind adding a test case for this too?

Comment (ignacio-chiazzo) [reply lib/irb.rb:976]: It is covered by https://github.com/ruby/irb/pull/851/files#diff-c37538faf5a6aa1413dc2844e6e78c58656b6474985c895e743ed422be4b3875R278-R280. I can change the message to something more explicit if needed.

File: lib/irb.rb [hunk 6]
```diff
@@ -979,13 +984,21 @@ def run(conf = IRB.conf)
...
           eval_input
         end
       ensure
         trap("SIGINT", prev_trap)
         conf[:AT_EXIT].each{|hook| hook.call}
-        context.io.save_history if save_history
+
+        if forced_exit
+          context.io.save_history if supports_history_saving
+          Kernel.exit!(0)
```
Comment (st0012) [inline lib/irb.rb:998]: I'm changing the method to `Kernel.exit` instead as it achieves the same effect we want but still runs `at_exit` callbacks.

I also don't think it'll be confusing to users as after the PR `exit!` will be listed in the help message as a command.

```suggestion
          Kernel.exit(0)
```

**Review by st0012 — APPROVED** [review]
Thanks for the PR ❤️

**Comment by nobu** [comment]
This added a new warning during the test.

[/home/runner/work/irb/irb/lib/irb/extend-command.rb:290: warning: delegator does not forward private method #exit!](https://github.com/ruby/irb/actions/runs/7857657441/job/21441803100#step:4:7)

**Pull request merged**